src/libfastrace.cpp

```diff
@@ -1,261 +1,436 @@
+      &fastrace_glue::ftr_create_def_coll_cfg);
+}
+
+ftr_coll_cfg ftr_set_max_spans_per_trace(ftr_coll_cfg cfg, size_t max) {
+  return call_rust_function<ftr_coll_cfg>(
+      &fastrace_glue::ftr_set_max_spans_per_trace,
+      *reinterpret_cast<ffi::ftr_coll_cfg*>(&cfg), max);
 }
 
 ftr_coll_cfg ftr_set_report_interval(ftr_coll_cfg cfg, uint64_t ri) {
-  union _ftr_coll_cfg _mcc = {
-      .f = fastrace_glue::ftr_set_report_interval(*(ffi::ftr_coll_cfg *)(&cfg),
-                                                  ri),
-  };
-  return std::move(_mcc.c);
+  return call_rust_function<ftr_coll_cfg>(
+      &fastrace_glue::ftr_set_report_interval,
+      *reinterpret_cast<ffi::ftr_coll_cfg*>(&cfg), ri);
 }
 
-void ftr_set_cons_rptr(void) { fastrace_glue::ftr_set_cons_rptr(); }
+void ftr_set_cons_rptr() { fastrace_glue::ftr_set_cons_rptr(); }
 
-ftr_otlp_exp_cfg ftr_create_def_otlp_exp_cfg(void) {
-  union _ftr_otlp_exp_cfg _moec = {
-      .f = fastrace_glue::ftr_create_def_otlp_exp_cfg(),
-  };
-  return std::move(_moec.c);
+ftr_otlp_exp_cfg ftr_create_def_otlp_exp_cfg() {
+  return call_rust_function<ftr_otlp_exp_cfg>(
+      &fastrace_glue::ftr_create_def_otlp_exp_cfg);
 }
 
 ftr_otel_rptr ftr_create_otel_rptr(ftr_otlp_exp_cfg cfg) {
-  union _ftr_otel_rptr _mor = {
-      .f =
-          fastrace_glue::ftr_create_otel_rptr(*(ffi::ftr_otlp_exp_cfg *)(&cfg)),
-  };
-  return std::move(_mor.c);
+  return call_rust_function<ftr_otel_rptr>(
+      &fastrace_glue::ftr_create_otel_rptr,
+      *reinterpret_cast<ffi::ftr_otlp_exp_cfg*>(&cfg));
 }
 
 void ftr_destroy_otel_rptr(ftr_otel_rptr rptr) {
-  fastrace_glue::ftr_destroy_otel_rptr(*(ffi::ftr_otel_rptr *)(&rptr));
+  fastrace_glue::ftr_destroy_otel_rptr(
+      *reinterpret_cast<ffi::ftr_otel_rptr*>(&rptr));
 }
 
 void ftr_set_otel_rptr(ftr_otel_rptr rptr, ftr_coll_cfg cfg) {
-  fastrace_glue::ftr_set_otel_rptr(*(ffi::ftr_otel_rptr *)(&rptr),
-                                   *(ffi::ftr_coll_cfg *)(&cfg));
+  fastrace_glue::ftr_set_otel_rptr(
+      *reinterpret_cast<ffi::ftr_otel_rptr*>(&rptr),
+      *reinterpret_cast<ffi::ftr_coll_cfg*>(&cfg));
 }
 
-void ftr_flush(void) { fastrace_glue::ftr_flush(); }
+void ftr_flush() { fastrace_glue::ftr_flush(); }
+
+}  // extern "C"
+
+namespace fastrace {
+
+SpanContext::SpanContext() : ctx_(ftr_create_rand_span_ctx()) {}
+
+SpanContext::SpanContext(const ftr_span_ctx& ctx) : ctx_(ctx) {}
+
+ftr_span_ctx SpanContext::raw() const { return ctx_; }
+
+void SpanContext::setSampled(bool sampled) {
+  ctx_ = ftr_span_ctx_set_sampled(ctx_, sampled);
 }
+
+Span::Span(const char* name, const SpanContext& parent)
+    : span_(new ftr_span(ftr_create_root_span(name, parent.raw()))) {}
+
+Span::Span(const char* name, const Span& parent)
+    : span_(new ftr_span(ftr_create_child_span_enter(name, parent.raw()))) {}
+
+Span::Span(const char* name)
+    : span_(new ftr_span(ftr_create_child_span_enter_loc(name))) {}
+
+Span::~Span() {
+  if (span_) {
```

Comment:
why `if(span_)`

Reply:
To support move constructor

Reply:
@andylokandy To support move constructor